drivers/xen/gnttab.c

```diff
@@ -0,0 +1,336 @@
+
+int gnttab_update_grant(grant_ref_t gref, domid_t domid,
+		unsigned long mfn, int readonly)
+{
+	int rc;
+
+	__ASSERT((gref >= GNTTAB_NR_RESERVED_ENTRIES &&
+		gref < NR_GRANT_ENTRIES), "Invalid gref = %d", gref);
+
+	rc = gnttab_reset_flags(gref);
+	if (!rc)
+		return rc;
+
+	gnttab_grant_permit_access(gref, domid, mfn, readonly);
+
+	return 1;
+}
+
+int gnttab_end_access(grant_ref_t gref)
+{
+	int rc;
+
+	__ASSERT((gref >= GNTTAB_NR_RESERVED_ENTRIES &&
+		gref < NR_GRANT_ENTRIES), "Invalid gref = %d", gref);
+
+	rc = gnttab_reset_flags(gref);
+	if (!rc) {
+		return rc;
+	}
+
+	put_free_entry(gref);
+
+	return 1;
+}
+
+unsigned long gnttab_end_transfer(grant_ref_t gref)
+{
+	unsigned long frame;
+	uint16_t flags;
+	uint16_t *pflags;
+
+	__ASSERT((gref >= GNTTAB_NR_RESERVED_ENTRIES &&
+		gref < NR_GRANT_ENTRIES), "Invalid gref = %d", gref);
+
+	pflags = &gnttab.table[gref].flags;
+	flags = *pflags;
+	while (!(flags & GTF_transfer_committed)) {
+		if (synch_cmpxchg(pflags, flags, 0) == flags) {
+			LOG_WRN("Release unused transfer grant.\n");
+			put_free_entry(gref);
+			return 0;
+		}
+		flags = *pflags;
+	}
+
+	/* If a transfer is in progress then wait until it is completed. */
+	while (!(flags & GTF_transfer_completed)) {
+		flags = *pflags;
+	}
+
+	/* Read the frame number /after/ reading completion status. */
+	compiler_barrier();
+	frame = gnttab.table[gref].frame;
+
+	put_free_entry(gref);
+
+	return frame;
+}
+
+grant_ref_t gnttab_alloc_and_grant(void **map)
+{
+	void *page;
+	unsigned long mfn;
+	grant_ref_t gref;
+
+	__ASSERT_NO_MSG(map != NULL);
+
+	page = k_aligned_alloc(XEN_PAGE_SIZE, XEN_PAGE_SIZE);
+	if (page == NULL) {
+		return -ENOMEM;
+	}
+
+	mfn = virt_to_mfn(page);
+	gref = gnttab_grant_access(0, mfn, 0);
+
+	*map = page;
+
+	return gref;
+}
+
+static void gop_eagain_retry(int cmd, struct gnttab_map_grant_ref *gref)
+{
+	unsigned int step = 10, delay = step;
+	int16_t *status = &gref->status;
+
+	do {
+		HYPERVISOR_grant_table_op(cmd, gref, 1);
+		if (*status == GNTST_eagain)
+			k_sleep(K_MSEC(delay));
+
+		delay += step;
+	} while ((*status == GNTST_eagain) && (delay < GOP_RETRY_DELAY));
+
+	if (delay >= GOP_RETRY_DELAY) {
+		LOG_ERR("Failed to map grant, timeout reached\n");
+		*status = GNTST_bad_page;
+	}
+}
+
+int gnttab_map_refs(struct gnttab_map_grant_ref *map_ops, unsigned int count)
+{
+	int i, ret;
+
+	ret = HYPERVISOR_grant_table_op(GNTTABOP_map_grant_ref, map_ops, count);
+	if (ret)
+		return ret;
+
+	for (i = 0; i < count; i++) {
+		switch (map_ops[i].status) {
+		case GNTST_no_device_space:
+			LOG_WRN("map_grant_ref failed, no device space for page #%d\n", i);
+			break;
+
+		case GNTST_eagain:
+			/* Operation not done; need to try again */
+			gop_eagain_retry(GNTTABOP_map_grant_ref, &map_ops[i]);
+			/* Need to re-check status for current page */
+			i--;
+
+			break;
+
+		default:
+			break;
+		}
+	}
+
+	return 0;
+}
+
+int gnttab_unmap_refs(struct gnttab_map_grant_ref *unmap_ops, unsigned int count)
```

Comment:
This call would leave a hole in the memory map. Should we document somewhere the address cannot be returned to the heap allocator afterwards?

Reply:
Thank you for very important advice, I didn't know about this and didn't test memory access after unmap - I confirmed that it become not accessible (I thought it only works it such way with remove_from_physmap).

In my opinion, left 4K holes it pretty small Zephyr heap memory poll is not a great idea, so I decided to fill them with populate_physmap at unmaped position. I tested it and it works fine, memory can be freed and reused, but I will be happy if you take a look at new implementation.

Reply:
The hypercall `GNTTABOP_map_grant_ref` will simply replace the entry in the P2M and not release any RAM that may have been associated.

So you would need to call `XENMEM_remove_from_physmap` before mapping the grant entry.  Otherwise, you will slowly exhaust the memory of the domain (note that this will not be a problem for dom0 because it is direct-mapped).

However, it is going to be quite expensive to unpopulate/populate every time you need to map/unmap a grant.

I think it would be better to consider to map grant entries outside of the actual RAM. Or, you unpopulate  part of the RAM (aka ballooning) for a longer period of time.